features/graph/api/src/main/java/org/opennms/netmgt/graph/api/generic/GenericEdge.java

```diff
@@ -39,24 +39,37 @@
 
 public class GenericEdge extends GenericElement implements Edge {
 
-    private final GenericVertex source;
-    private final GenericVertex target;
+    private final GenericVertexRef source;
+    private final GenericVertexRef target;
 
-    public GenericEdge(GenericVertex source, GenericVertex target) {
-        this(source, target, new HashMap<>());
+    public GenericEdge(String namespace, GenericVertexRef source, GenericVertexRef target) {
+        this(source, target, createMapWithNamespace(namespace));
     }
 
-    public GenericEdge(GenericVertex source, GenericVertex target, Map<String, Object> properties) {
+    private static Map<String, Object> createMapWithNamespace(String namespace) {
+        Objects.requireNonNull(namespace);
+        Map<String, Object> properties = new HashMap<>();
+        properties.put(GenericProperties.NAMESPACE, namespace);
+        return properties;
+    }
+
+    public GenericEdge(GenericVertexRef source, GenericVertexRef target, Map<String, Object> properties) {
         super(properties);
         this.source = Objects.requireNonNull(source);
         this.target = Objects.requireNonNull(target);
-        this.setNamespace(source.getNamespace());
-        this.setId(source.getId() + "->" + target.getId());
+        // TODO patrick: discuss with mvr: can an edge connect the same vertexref? If not we want to add a check?
+        if(!source.getNamespace().equals(getNamespace()) && !target.getNamespace().equals(getNamespace())) {
```

Comment:
This is in general a bit tricky.
This check implies, that the namespace is contained in the properties.
Is this really always the case?
The namespace should at least be enforced before trying to equal on it

Besides this, an edge can connect the same vertex, so pointing to itself. 
This may not be used often, but AFAIK it can already happen within linkd.

Reply:
I made sure that a GenericGraph, GenericVertex, GenericVertexRef and Edge always contain a non null namespace. This check is done in GenericElement: https://github.com/OpenNMS/opennms/blob/58756b55bc9e97ee619ec6479a958e41b48d9564/features/graph/api/src/main/java/org/opennms/netmgt/graph/api/generic/GenericElement.java#L48